components/raftstore/src/store/async_io/write.rs

```diff
@@ -0,0 +1,839 @@
+                    }
+                };
+                match msg {
+                    SyncMsg::Shutdown => return,
+                    SyncMsg::Sync {
+                        worker_id,
+                        unpersisted_id,
+                    } => {
+                        persist_vec[worker_id] = Some(unpersisted_id);
+                    }
+                }
+            }
+
+            let now = Instant::now();
+            self.sync.sync();
+            STORE_WRITE_SYNC_RAFT_DB_DURATION_HISTOGRAM
+                .observe(duration_to_sec(now.saturating_elapsed()));
+
+            for i in 0..persist_vec.len() {
+                if let Some(persisted_id) = persist_vec[i] {
+                    self.global_persisted_ids[i].store(persisted_id, Ordering::Release);
+                    // Sync worker is destroyed after write worker
+                    let _ = self.write_senders[i].send(WriteMsg::Persisted);
+                    persist_vec[i] = None;
+                }
+            }
+        }
+    }
+}
+
+pub struct StoreWriters<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    writers: Vec<Sender<WriteMsg<EK, ER>>>,
+    sync_sender: Option<Sender<SyncMsg>>,
+    handlers: Vec<JoinHandle<()>>,
+    sync_handler: Option<JoinHandle<()>>,
+}
+
+impl<EK, ER> StoreWriters<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    pub fn new() -> Self {
+        Self {
+            writers: vec![],
+            sync_sender: None,
+            handlers: vec![],
+            sync_handler: None,
+        }
+    }
+
+    pub fn senders(&self) -> &Vec<Sender<WriteMsg<EK, ER>>> {
+        &self.writers
+    }
+
+    pub fn spawn<T: Transport + 'static, N: Notifier>(
+        &mut self,
+        store_id: u64,
+        engines: &Engines<EK, ER>,
+        notifier: &N,
+        trans: &T,
+        cfg: &Arc<VersionTrack<Config>>,
+    ) -> Result<()> {
+        let (sync_sender, sync_receiver) = unbounded();
+        let mut global_persisted_ids = vec![];
+        let pool_size = cfg.value().store_io_pool_size;
+        for i in 0..pool_size {
+            let tag = format!("store-writer-{}", i);
+            let (tx, rx) = bounded(cfg.value().store_io_notify_capacity);
+            let global_persisted_id = Arc::new(CachePadded::new(AtomicU64::new(0)));
+            let mut worker = Worker::new(
+                store_id,
+                tag.clone(),
+                i,
+                engines.clone(),
+                rx,
+                notifier.clone(),
+                trans.clone(),
+                global_persisted_id.clone(),
+                sync_sender.clone(),
+                cfg,
+            );
+            let t = thread::Builder::new().name(thd_name!(tag)).spawn(move || {
+                worker.run();
+            })?;
+            global_persisted_ids.push(global_persisted_id);
+            self.writers.push(tx);
+            self.handlers.push(t);
+        }
+        let mut syncer = SyncWorker::new(
```

Comment:
What are the benefits to add an extra `SyncWorker`? IMO there is no difference to queueing write task in `channel` or  `Worker:: pending_tasks` and having a single sync worker and multiple write worker may cause a task need to wait for all previous tasks from all worker to finish before start.

Reply:
There is only one `fsync` at the same time, which is the same as no `SyncWorker`. So there is no extra wait time. 
The benefit is that the multiple `Worker` can do more cpu workers like `batch WriteTask`, `send callback to peer` and `send msg` instead of waiting for `fsync`.

Reply:
What I mean is that it may trigger more `fsync` than before because the `WriteWorker` has less time of blocking io, the batch size may reduce hence trigger more `SyncMsg::Sync` and wait more `fsync`.

Also, we may wait for more blocking io operations to send messages to other peers than before:
- Without `SyncWorker`: `prev_write` + `prev_fsync` + `write` + `fsync` -> send msg
- Wirth `SyncWorker`: `prev_write` + `write` + `prev_fsync` + `fsync` + `later_coming_write` -> send msg